File: CppLibrary/Collection/kosaraju.cpp
#include <iostream>
#include <cstring>
using namespace std;
const int maxn = 2000;
const int maxedge = 10000 + 1;
struct edgenode
{
	int v, next;
};

edgenode edge[maxedge], inv_edge[maxedge];
int edgenum = 0, inv_edgenum = 0;
int start[maxn], inv_start[maxn];
void insert(int x, int y)
{
	edge[++ edgenum].v = y;
	edge[edgenum].next = start[x];
	start[x] = edgenum;
	inv_edge[++ inv_edgenum].v = x;
	inv_edge[inv_edgenum].next = inv_start[y];
	inv_start[y] = inv_edgenum;
}

int n;
int order[maxn];

void init()
{
	memset(start, -1, sizeof(start));
	memset(inv_start, -1, sizeof(inv_start));
	int m;
	cin >> n >> m;
	int x, y;
	for (int i = 0; i < m; ++ i)
	{
		cin >> x >> y;
		insert(x, y);
	}
}

int ordernum = 0, totalscc = 0;
bool vis[maxn];
int flag[maxn], scc[maxn];

void dfs(int x)
{
	vis[x] = true;
	for (int i = start[x]; i >= 0; i = edge[i].next)
		if (!vis[edge[i].v]) dfs(edge[i].v);
	order[++ ordernum] = x;
}

void inv_dfs(int x)
{
	vis[x] = true; flag[x] = totalscc;
	for (int i = inv_start[x]; i >= 0; i = inv_edge[i].next)
		if (!vis[inv_edge[i].v]) inv_dfs(inv_edge[i].v);
}

void work()
{
	memset(flag, 0, sizeof(flag));
	memset(vis, false, sizeof(vis));
	for (int i = 0; i < n; ++ i)
		if (!vis[i]) dfs(i);
	memset(vis, false, sizeof(vis));
	for (int i = ordernum; i >= 1; -- i)
		if (!vis[order[i]])
		{
			++ totalscc;
			inv_dfs(order[i]);
		}
	memset(scc, 0, sizeof(scc));
	
	for (int i = 0; i < n; ++ i)
		++ scc[flag[i]];
		
	// Check each point whether in cycle
	// Attention the case: only one point construct a cycle
	for (int i = 0; i < n; ++ i)
	{
		if (scc[flag[i]] > 1)
			cout << "Yes" << endl;
		else
			cout << "No" << endl;
	}
}

int main()
{
	//freopen("in.txt", "r", stdin);
	init();
	work();
	return 0;
}

File: PascalLibrary/Collection/kosaraju.pas
program kosaraju;
const
	maxn = 1000;
	maxm = 10000;
type
	edgenode = record
		next,v:longint;
	end;
var
	path,inpath:array[1..maxm] of edgenode;
	starta,startb:array[1..maxn] of longint;
	order,flag,scc:array[1..maxn] of longint;
	vis:array[1..maxn] of boolean;
	edgenum,n,totalscc,ordernum:longint;
	
procedure insert(x,y:longint);
 begin
	inc(edgenum);
	path[edgenum].v:=y;
	path[edgenum].next:=starta[x];
	starta[x]:=edgenum;
	inpath[edgenum].v:=x;
	inpath[edgenum].next:=startb[y];
	startb[y]:=edgenum;
 end; 

procedure init;
 var
	i,m,x,y:longint;
 begin
	readln(n,m);
	edgenum:=0;
	for i:=1 to m do begin
		readln(x,y);
		insert(x,y);
	end;
 end;

procedure dfs(x:longint);
 var
	p:longint;
 begin
	vis[x]:=true;
	p:=starta[x];
	while (p > 0) do begin
		if (not vis[path[p].v]) then dfs(path[p].v);
		p:=path[p].next;
	end;
	inc(ordernum);
	order[ordernum]:=x;
 end;
 
procedure redfs(x:longint);
 var
	p:longint;
 begin
	vis[x]:=true; flag[x]:=totalscc;
	p:=startb[x];
	while (p > 0) do begin
		if (not vis[inpath[p].v]) then redfs(inpath[p].v);
		p:=inpath[p].next;
	end;
 end;
 
procedure main;
 var
	i:longint;
 begin
	ordernum:=0;
	fillchar(vis,sizeof(vis),false);
	for i:=1 to n do
		if (not vis[i]) then dfs(i);
	fillchar(vis,sizeof(vis),false);
	totalscc:=0;
	for i:=ordernum downto 1 do
		if (not vis[order[i]]) then begin
			inc(totalscc);
			redfs(order[i]);
		end;
	fillchar(scc,sizeof(scc),0);
	for i:=1 to n do inc(scc[flag[i]]);
 end;
 
procedure print;
 var
	i:longint;
 begin
	writeln('Totalscc = ',totalscc);
	for i:=1 to n do writeln(i,'-->',flag[i]);
	writeln('Count ---->');
	for i:=1 to totalscc do writeln(i,'-->',scc[i]);
 end;
 
begin
	init;
	main;
	print;
end.
